riscv/src/asm_translate/mod.rs

```diff
@@ -0,0 +1,575 @@
+    fn rr(&self) -> Result<(Register, Register), &'static str> {
+        match self {
+            [Argument::Register(r1), Argument::Register(r2)] => Ok((*r1, *r2)),
+            _ => Err("Expected: register, register"),
+        }
+    }
+
+    fn rrl(&self) -> Result<(Register, Register, String), &'static str> {
+        const ERR: &str = "Expected: register, register, label";
+        match self {
+            [Argument::Register(r1), Argument::Register(r2), l] => {
+                Ok((*r1, *r2, argument_to_escaped_symbol(l).ok_or(ERR)?))
+            }
+            _ => Err(ERR),
+        }
+    }
+
+    fn rl(&self) -> Result<(Register, String), &'static str> {
+        const ERR: &str = "Expected: register, label";
+        match self {
+            [Argument::Register(r1), l] => Ok((*r1, argument_to_escaped_symbol(l).ok_or(ERR)?)),
+            _ => Err(ERR),
+        }
+    }
+
+    fn rro(&self) -> Result<(Register, Register, u32), &'static str> {
+        if let [Argument::Register(r1), Argument::RegOffset(off, r2)] = self {
+            if let Some(off) = expression_to_number(off.as_ref().unwrap_or(&Expression::Number(0)))
+            {
+                return Ok((*r1, *r2, off));
+            }
+        }
+        if let [Argument::Register(r1), Argument::Expression(off)] = self {
+            if let Some(off) = expression_to_number(off) {
+                // If the register is not specified, it defaults to x0
+                return Ok((*r1, Register::new(0), off));
+            }
+        }
+
+        Err("Expected: register, offset(register)")
+    }
+
+    fn rrro(&self) -> Result<(Register, Register, Register, u32), &'static str> {
+        if let [Argument::Register(r1), Argument::Register(r2), Argument::RegOffset(off, r3)] = self
+        {
+            if let Some(off) = expression_to_number(off.as_ref().unwrap_or(&Expression::Number(0)))
+            {
+                return Ok((*r1, *r2, *r3, off));
+            }
+        }
+        if let [Argument::Register(r1), Argument::Register(r2), Argument::Expression(off)] = self {
+            if let Some(off) = expression_to_number(off) {
+                // If the register is not specified, it defaults to x0
+                return Ok((*r1, *r2, Register::new(0), off));
+            }
+        }
+        Err("Expected: register, register, offset(register)")
+    }
+
+    fn empty(&self) -> Result<(), &'static str> {
+        match self {
+            [] => Ok(()),
+            _ => Err("Expected: no arguments"),
+        }
+    }
+}
+
+/// Compiles riscv assembly to a powdr assembly file. Adds required library routines.
+pub fn compile<F: FieldElement>(
+    assemblies: BTreeMap<String, String>,
+    runtime: &Runtime,
+    with_bootloader: bool,
+) -> String {
+    let asm_program = compile_internal(assemblies);
+
+    code_gen::translate_program::<F>(&asm_program, runtime, with_bootloader)
+}
+
+fn compile_internal(mut assemblies: BTreeMap<String, String>) -> AsmProgram {
```

Comment:
maybe `compile_asm`?

Reply:
This is just so that the big function don't have generic parameters (so that the generated code is smaller). So I think this function should be name `<name_of_the_public_function>_internal`